tensorflow_model_optimization/python/core/sparsity_tf2/lthpruner_test.py

```diff
@@ -0,0 +1,486 @@
+    self.assertAllEqual(np.count_nonzero(mask_after_pruning), 97)
+
+  
+  def testReloadWeightsIterationK(self):
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+
+    save_round = 5
+    n_rounds = 24
+    end_iter = 100
+    frequency, prune_ratio_per_round = get_lth_sparsity(save_round, n_rounds, self.target_sparsity, end_iter)
+    pruning_schedule = make_pruning_schedule(1 - prune_ratio_per_round, save_round + 1, end_iter, frequency)
+    
+    p = pruner.LTHPruner(
+        pruning_schedule=pruning_schedule,
+        save_iteration=save_round,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+
+    def _train(weight):
+      p.create_slots(optimizer, weight)
+      for i in range(7):
+        p.preprocess_weights(optimizer, weight, self.grad(weight))
+        weight.assign(tf.math.add(weight, sample_noise(i)))
+        p.postprocess_weights(optimizer, weight, self.grad(weight))
+        optimizer.iterations.assign_add(1)
+
+    _train(weight)
+    masked_orig_init = tf.math.multiply(optimizer.get_slot(weight, "original_initialization"), optimizer.get_slot(weight, 'mask'))
+    self.assertAllEqual(masked_orig_init, weight)
+
+    mask_after_pruning = optimizer.get_slot(weight, "mask").read_value()
+    self.assertAllEqual(np.count_nonzero(mask_after_pruning), 97)
+
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+    optimizer.iterations.assign(0)
+    tf.function(_train)(weight)
+    masked_orig_init = tf.math.multiply(optimizer.get_slot(weight, "original_initialization"), optimizer.get_slot(weight, 'mask'))
+    self.assertAllEqual(masked_orig_init, weight)
+
+    mask_after_pruning = optimizer.get_slot(weight, "mask").read_value()
+    self.assertAllEqual(np.count_nonzero(mask_after_pruning), 97)
+
+
+  def testReloadTwoTimes(self):
+    weight = tf.Variable(np.linspace(1.0, 100.0, 100), name="weights")
+    weight_dtype = weight.dtype.base_dtype
+
+    save_round = 5
+    n_rounds = 24
+    end_iter = 100
+    frequency, prune_ratio_per_round = get_lth_sparsity(save_round, n_rounds, self.target_sparsity, end_iter)
+    pruning_schedule = make_pruning_schedule(1 - prune_ratio_per_round, save_round + 1, end_iter, frequency)
+    
+    p = pruner.LTHPruner(
+        pruning_schedule=pruning_schedule,
+        save_iteration=save_round,
+        block_size=self.block_size,
+        block_pooling_type=self.block_pooling_type)
+
+    optimizer = self.dummy_optimizer
+    optimizer.iterations.assign(0)
+
+    def _train(weight):
+      expected_first_saved_initialization = None
+      expected_second_saved_initialization = None
+      p.create_slots(optimizer, weight)
+      for i in range(0, 13): # this should save, reload, and update exactly twice
+        p.preprocess_weights(optimizer, weight, self.grad(weight))
+        if i == save_round - 1:
```

Comment:
Let's capture the weight after post_process weights (when it will be reloaded + masked).

(assuming this is the correct step to check against? it might not be?)